### PR TITLE
thread: allow specifying stack size for new thread

### PR DESCRIPTION
**Pull request by addaleax**

### CODE REVIEW

**Comment by addaleax** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1198/

**Comment by refack** [comment]
Code LGTM in general, but if we're venturing into `*_ex` territory, I'd rather we plan ahead and do it by defining a `uv_thread_ex_t`:
```c
struct uv_thread_ex_s {
  uv_thread_t fd;
  uv_thread_cb entry;
  void* arg;
  size_t stack_size;
  /* private */
  void* reserved[6];
};
typedef uv_thread_ex_t uv_thread_ex_s;
```

**Comment by addaleax** [comment]
@refack Comparing `_beginthreadex` and pthread attributes, this seems to be the only overlap we could forward in a cross-platform manner.

**Comment by addaleax** [comment]
(I’m up for implementing this based on a struct if there’s more things we could see us extending this to.)

CI: https://ci.nodejs.org/job/libuv-test-commit/1201/

**Comment by refack** [comment]
> I’m up for implementing this based on a struct if there’s more things we could see us extending this to.

Mostly trying to think about forward compatibility, and assuming OS level APIs might progress. I'd assume things like; start state, CPU affinity, security, locale, name, or simply stuff we want to put in TLS...

Anyway, not a hard -1, just a suggestion. Happy to hear more opinions!

**Comment by addaleax** [comment]
@refack I’ve pushed a version that would use a struct rather than an individual extra parameter – do you like that better?

I don’t like the extra complexity for the documentation but other than that it seems okay to me.

**Comment by saghul** [comment]
Me likey! FWIW we do have affinity and detaching as functions in master. Since v2 is not released maybe folding that functionality into this structure could be considered in the future. Nice work @addaleax !

Suggestion: maybe add some reserved fields for good meassure?

**Comment by addaleax** [comment]
@saghul As long as we’re only consuming the struct and access to fields is only done when indicated by flags, we can always extend the struct ourselves when adding new features, right?

**Comment by saghul** [comment]
@addaleax That would break the ABI.

**Comment by addaleax** [comment]
@saghul It wouldn’t – unless users do much fancier stuff with the options struct that you wouldn’t expect them to do, like expose it as part of their own ABI or so. I can add a note about the possibility of extra future fields, if you want.

**Comment by addaleax** [comment]
@saghul Does 83b148676a18ccb85a528e58929edaa2ec29981e sound okay to you?

**Comment by saghul** [comment]
Please correct me if I'm wrong, but I was under the impression that modifying a (non-internal) struct layout implies ABI brekage. That's why we added the 4 reserved fields on handles, so we could "change" it without actually doing so, if need be. Or am I not caffeinated enough?

**Comment by addaleax** [comment]
@saghul It depends on which side of the API the struct is allocated and on which side it is read/written to, and whether things like subclassing play a role here (like they do for handles).

In this case, we only read the struct’s fields, so the only danger would be accessing field that do not exist in the user’s code. But we only do so when we know that they are present (because the user has set the corresponding flag), so that’s okay.

We could return `UV_EINVAL` if we see unknown flags, i.e. flags/fields added in a later libuv version, rather than ignoring them, if you think that that’s a good idea.

**Comment by addaleax** [comment]
@refack @saghul Any further thoughts? I don’t want to rush this but it would be nice to have for Workers in Node :)

File: src/unix/thread.c [hunk 1]
```diff
@@ -194,13 +194,34 @@ static size_t thread_stack_size(void) {
...
+      params->flags & UV_THREAD_HAS_STACK_SIZE ? params->stack_size : 0;
 
   attr = NULL;
-  stack_size = thread_stack_size();
+  if (stack_size == 0) {
+    stack_size = thread_stack_size();
+  } else {
+    pagesize = getpagesize();
+    if (stack_size % pagesize != 0)
+      stack_size += pagesize - (stack_size % pagesize);
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Not exactly wrong but how about this?
```c
stack_size += pagesize - 1;
stack_size &= ~(size_t)pagesize;
```

**Comment by addaleax** [comment]
@bnoordhuis Thanks for the review, should be addressed in db0a26d061d35f89076575038d1500a929d1b357

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1208/

File: src/unix/thread.c [hunk 2]
```diff
@@ -216,9 +216,9 @@ int uv_thread_create_ex(uv_thread_t* tid,
   if (stack_size == 0) {
     stack_size = thread_stack_size();
   } else {
-    pagesize = getpagesize();
-    if (stack_size % pagesize != 0)
-      stack_size += pagesize - (stack_size % pagesize);
+    pagesize = (size_t)getpagesize();
+    // Round up to the nearest page boundary.
```
Comment (bnoordhuis) [inline src/unix/thread.c]: No C99/C++ style comments.

Comment (addaleax) [reply src/unix/thread.c]: Sorry, done!

**Comment by addaleax** [comment]
Landed in 0eca049a9b3ca7440b4a4c00ab0ebe57cc3ff948, thanks for the reviews!

**Pull request closed without merging**